src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs

```diff
@@ -69,22 +70,43 @@ public Task OnConnectionAsync(ConnectionContext connectionContext)
             }
 
             var connection = new HttpConnection(httpConnectionContext);
+            var inputCompletionState = new PipeCompletionState(connection);
+            var outputCompletionState = new PipeCompletionState(connection);
 
             var processingTask = connection.StartRequestProcessing(_application);
 
-            connectionContext.Transport.Input.OnWriterCompleted((error, state) =>
-            {
-                ((HttpConnection)state).Abort(error);
-            },
-            connection);
+            connectionContext.Transport.Input.OnWriterCompleted(
+                (error, state) => CompletedCallback(error, state),
+                inputCompletionState);
+
+            connectionContext.Transport.Output.OnReaderCompleted(
+                (error, state) => CompletedCallback(error, state),
+                outputCompletionState);
+
+            await inputCompletionState.CompletedTcs.Task;
+            await outputCompletionState.CompletedTcs.Task;
+
+            connection.OnConnectionClosed();
 
-            connectionContext.Transport.Output.OnReaderCompleted((error, state) =>
+            await processingTask;
+        }
+
+        private static void CompletedCallback(Exception error, object state)
+        {
+            var pipeCompletionState = (PipeCompletionState)state;
+            pipeCompletionState.Connection.Abort(error);
+            pipeCompletionState.CompletedTcs.SetResult(null);
+        }
+
+        private class PipeCompletionState
+        {
+            public PipeCompletionState(HttpConnection connection)
             {
-                ((HttpConnection)state).OnConnectionClosed(error);
-            },
-            connection);
+                Connection = connection;
+            }
 
-            return processingTask;
+            public HttpConnection Connection;
+            public TaskCompletionSource<object> CompletedTcs = new TaskCompletionSource<object>();
```

Comment:
Do you want to run these continuations inline?

Reply:
The pipe completion callbacks already get dispatched by default. This is closer to calling Abort and OnConnectionClosed directly in the callback like before.